LBH.AdultSocialCare.Api.Tests/V1/Gateways/Common/CarePackageHistoryGatewayTests.cs

```diff
@@ -25,37 +27,29 @@ public CarePackageHistoryGatewayTests()
         public async Task ShouldReturnHistoriesForPackage()
         {
             //Arrange
-            var user = new Client() { Id = Guid.Parse(UserConstants.DefaultApiUserId) };
-            Context.Clients.Add(user);
-            await Context.SaveChangesAsync().ConfigureAwait(false);
-
-            var carePackage = new CarePackage() { ServiceUserId = Guid.Parse(UserConstants.DefaultApiUserId) };
-            Context.CarePackages.Add(carePackage);
-            await Context.SaveChangesAsync().ConfigureAwait(false);
+            var carePackageId = new Guid();
+            var historyList = new List<CarePackageHistory>();
 
-            var carePackageId = carePackage.Id;
-            var carePackageHistory = new List<CarePackageHistory>()
+            for (var i = 0; i < 10; i++)
             {
-                new CarePackageHistory
-                {
-                    CarePackageId = carePackageId,
-                    Description = "test"
-                },
-                new CarePackageHistory
-                {
-                    CarePackageId = carePackageId,
-                    Description = "test 1"
-                },
-            };
+                historyList.Add(TestDataHelper.CreateCarePackageHistory(carePackageId: carePackageId));
+            }
 
-            Context.CarePackageHistories.AddRange(carePackageHistory);
-            await Context.SaveChangesAsync().ConfigureAwait(false);
+            var carePackageHistory = historyList.AsAsyncQueryable();
+
+            var mockSet = new Mock<DbSet<CarePackageHistory>>();
+            mockSet.As<IQueryable<CarePackageHistory>>().Setup(m => m.Provider).Returns(carePackageHistory.Provider);
+            mockSet.As<IQueryable<CarePackageHistory>>().Setup(m => m.Expression).Returns(carePackageHistory.Expression);
+            mockSet.As<IQueryable<CarePackageHistory>>().Setup(m => m.ElementType).Returns(carePackageHistory.ElementType);
+            mockSet.As<IQueryable<CarePackageHistory>>().Setup(m => m.GetEnumerator()).Returns(carePackageHistory.GetEnumerator());
+
+            Context.CarePackageHistories = mockSet.Object;
```

Comment:
No need to mock DataSets of already configured Context. In BaseInMemoryTest Context is configured for SQLite. Context initialization is the slowest part and it already happened. Adding / reading data is relatively cheap.

As general rule, concentrate on 
- E2E tests to ensure that everything (DI, routing, mapping, DB) is wired up correctly
- UseCase and related component tests to validate different aspects of business logic

Setup for gateway tests takes ~2-3 sec, unlike normal test which takes ~0.5 sec. So there is not too much reason in testing simple gateways (it will be tested by E2E anyway), but it still can be reasonable to test complex queries

Reply:
Here is the thinking, Setup process to do this test is complex. Many foreign keys  I need to setup. Why do that when I can mock the dataset leave the FK thing for the E2E test. Fair point though on putting more focus on unit testing use cases and E2E

Reply:
By setup I mean BaseInMemoryTest constructor, which creates database and takes 2-3 seconds

We use mocks to imitate external dependencies, but in this case mock is used to replace real implementation, so test tests mock, and not implementation.

So, I wouldn't create this test, becasue it should be covered by E2E anyway.

If you still believe that this test is necessary, then setup foreign keys to create real test scenario - as test should test exactly real usage with all dependencies in place.

If you still believe that testing mock make sense, move it away from BaseInMemoryTest, since this extra 2-3 sec of setup time aren't needed there